lib/kino/smart_cell/chart_builder.ex

```diff
@@ -0,0 +1,210 @@
+defmodule Kino.SmartCell.ChartBuilder do
+  use Kino.JS, assets_path: "lib/assets/chart_builder"
+  use Kino.JS.Live
+  use Kino.SmartCell, name: "Chart builder"
+
+  @as_int ["width", "height"]
+  @as_atom ["data_variable", "chart_type", "x_field_type", "y_field_type", "color_field_type"]
+
+  @impl true
+  def init(attrs, ctx) do
+    fields = %{
+      "chart_type" => attrs["chart_type"] || "bar",
+      "width" => attrs["width"],
+      "height" => attrs["height"],
+      "x_field" => attrs["x_field"],
+      "y_field" => attrs["y_field"],
+      "x_field_type" => attrs["x_field_type"],
+      "y_field_type" => attrs["y_field_type"],
+      "color_field" => attrs["color_field"],
+      "color_field_type" => attrs["color_field_type"],
+      "data_variable" => attrs["data_variable"]
+    }
+
+    ctx =
+      assign(ctx,
+        fields: fields,
+        options: %{},
+        vl_alias: nil,
+        missing_dep: missing_dep(),
+        fresh: attrs == %{}
+      )
+
+    {:ok, ctx, reevaluate_on_change: true}
+  end
+
+  @impl true
+  def scan_binding(pid, binding, env) do
+    data_options =
+      for {key, val} <- binding, is_atom(key), is_map(val), into: %{}, do: {key, Map.keys(val)}
+
+    vl_alias = vl_alias(env)
+    send(pid, {:scan_binding_result, data_options, vl_alias})
+  end
+
+  @impl true
+  def handle_connect(ctx) do
+    payload = %{
+      fields: ctx.assigns.fields,
+      missing_dep: ctx.assigns.missing_dep,
+      options: ctx.assigns.options,
+      fresh: ctx.assigns.fresh
+    }
+
+    {:ok, payload, ctx}
+  end
+
+  @impl true
+  def handle_info({:scan_binding_result, data_options, vl_alias}, ctx) do
+    ctx = assign(ctx, options: data_options, vl_alias: vl_alias)
+    broadcast_event(ctx, "set_available_data", %{"options" => data_options})
+
+    {:noreply, ctx}
+  end
+
+  @impl true
+  def handle_event("update_field", %{"field" => "data_variable", "value" => ""}, ctx) do
```

Comment:
Shouldn't we reset other fields in this case? We can match in `update_data_options` on empty value, and return all empty fields :)

Reply:
Nops! Here we have a specific case. `data_variable` will be `""` if and only if the ChartBuilder has never been populated before. So we have nothing to reset and we don't need to broadcast.

Reply:
This happens in two scenarios: when a ChartBuilder is started without variables and when reconnection with an empty (or fresh) ChartBuilder occurs.

Reply:
What I meant is that if it's the JS side triggering and we don't want it, we can always add a check like

```js
if (name === "data_variable" && value === "") { return }
```